risc0/zkvm/src/host/server/prove/prover_impl.rs

```diff
@@ -132,60 +124,24 @@ where
     }
 
     fn prove_segment(&self, ctx: &VerifierContext, segment: &Segment) -> Result<SegmentReceipt> {
-        use risc0_zkp::prove::executor::Executor;
-
-        tracing::debug!(
-            "prove_segment[{}]: po2: {}, cycles: {}",
-            segment.index,
-            segment.po2,
-            segment.cycles,
-        );
-        let (hal, circuit_hal) = (self.hal_pair.hal.as_ref(), &self.hal_pair.circuit_hal);
-        let hashfn = &hal.get_hash_suite().name;
-
-        let io = segment.prepare_globals()?;
-        let machine = MachineContext::new(segment);
-        let po2 = segment.po2 as usize;
-        let mut executor = Executor::new(&CIRCUIT, machine, po2, po2, &io);
+        use risc0_circuit_rv32im::prove::{engine::SegmentProverImpl, SegmentProver as _};
 
-        let loader = Loader::new();
-        loader.load(|chunk, fini| executor.step(chunk, fini))?;
-        executor.finalize();
-
-        let mut adapter = ProveAdapter::new(&mut executor);
-        let mut prover = risc0_zkp::prove::Prover::new(hal, CIRCUIT.get_taps());
-
-        adapter.execute(prover.iop());
-
-        prover.set_po2(adapter.po2() as usize);
-
-        prover.commit_group(
-            REGISTER_GROUP_CODE,
-            hal.copy_from_elem("code", &adapter.get_code().as_slice()),
-        );
-        prover.commit_group(
-            REGISTER_GROUP_DATA,
-            hal.copy_from_elem("data", &adapter.get_data().as_slice()),
-        );
-        adapter.accumulate(prover.iop());
-        prover.commit_group(
-            REGISTER_GROUP_ACCUM,
-            hal.copy_from_elem("accum", &adapter.get_accum().as_slice()),
-        );
+        use crate::host::receipt::decode_receipt_claim_from_seal;
 
-        let mix = hal.copy_from_elem("mix", &adapter.get_mix().as_slice());
-        let out_slice = &adapter.get_io().as_slice();
+        let hashfn = self.hal_pair.hal.get_hash_suite().name.clone();
 
-        tracing::debug!("Globals: {:?}", OutBuffer(out_slice).tree(LAYOUT));
-        let out = hal.copy_from_elem("out", &adapter.get_io().as_slice());
+        let prover =
+            SegmentProverImpl::new(self.hal_pair.hal.clone(), self.hal_pair.circuit_hal.clone());
+        let seal = prover.prove_segment(&segment.inner)?;
 
-        let seal = prover.finalize(&[&mix, &out], circuit_hal.as_ref());
+        let mut claim = decode_receipt_claim_from_seal(&seal)?;
+        claim.output = segment.output.clone().into();
 
         let receipt = SegmentReceipt {
             seal,
-            index: segment.index,
-            hashfn: hashfn.clone(),
-            claim: segment.get_claim()?,
+            index: segment.index as u32,
+            hashfn,
+            claim,
```

Comment:
@nategraf Note that instead of creating a new claim from scratch, this impl will decode the one from the seal and then inject the output (to un-prune it).